File: cloud/internal/middleware/session.go
package middleware

import (
	"context"
	"net/http"

	"github.com/google/uuid"
	"github.com/kx0101/replayer-cloud/internal/auth"
	"github.com/kx0101/replayer-cloud/internal/models"
	"github.com/kx0101/replayer-cloud/internal/store"
)

type contextKey string

const (
	SessionContextKey contextKey = "session"
	UserContextKey    contextKey = "user"
	UserIDContextKey  contextKey = "user_id"
)

func SessionAuth(sm *auth.SessionManager, s store.Store) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			session, err := sm.GetSession(r)
			if err != nil {
				http.Redirect(w, r, "/login", http.StatusSeeOther)
				return
			}

			ctx := context.WithValue(r.Context(), SessionContextKey, session)
			ctx = context.WithValue(ctx, UserIDContextKey, session.UserID)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

func OptionalSession(sm *auth.SessionManager) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			session, err := sm.GetSession(r)
			if err == nil {
				ctx := context.WithValue(r.Context(), SessionContextKey, session)
				ctx = context.WithValue(ctx, UserIDContextKey, session.UserID)
				r = r.WithContext(ctx)
			}
			next.ServeHTTP(w, r)
		})
	}
}

func RequireVerified(sm *auth.SessionManager, s store.Store) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			session, err := sm.GetSession(r)
			if err != nil {
				http.Redirect(w, r, "/login", http.StatusSeeOther)
				return
			}

			user, err := s.GetUserByID(r.Context(), session.UserID)
			if err != nil || user == nil {
				sm.ClearSession(w)
				http.Redirect(w, r, "/login", http.StatusSeeOther)
				return
			}

			if user.VerifiedAt == nil {
				http.Redirect(w, r, "/verify-pending", http.StatusSeeOther)
				return
			}

			ctx := context.WithValue(r.Context(), SessionContextKey, session)
			ctx = context.WithValue(ctx, UserContextKey, user)
			ctx = context.WithValue(ctx, UserIDContextKey, session.UserID)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

func GetSession(ctx context.Context) *models.SessionData {
	if session, ok := ctx.Value(SessionContextKey).(*models.SessionData); ok {
		return session
	}
	return nil
}

func GetUser(ctx context.Context) *models.User {
	if user, ok := ctx.Value(UserContextKey).(*models.User); ok {
		return user
	}
	return nil
}

func GetUserID(ctx context.Context) uuid.UUID {
	if id, ok := ctx.Value(UserIDContextKey).(uuid.UUID); ok {
		return id
	}
	return uuid.Nil
}

File: crates/replayer-cloud/src/middleware/session.rs
use axum::{
    body::Body,
    extract::State,
    http::Request,
    middleware::Next,
    response::{IntoResponse, Redirect, Response},
};

use crate::handler::AppState;
use crate::models::{SessionData, User};

use super::apikey::AuthUserId;

#[derive(Clone, Debug)]
pub struct AuthSession(#[allow(dead_code)] pub SessionData);

#[derive(Clone, Debug)]
pub struct AuthUser(pub User);

pub async fn optional_session(
    State(state): State<AppState>,
    mut req: Request<Body>,
    next: Next,
) -> Response {
    let cookie_header = req.headers().get("cookie").and_then(|v| v.to_str().ok());

    if let Ok(session) = state.session_manager.get_session(cookie_header) {
        req.extensions_mut().insert(AuthUserId(session.user_id));
        req.extensions_mut().insert(AuthSession(session));
    }

    next.run(req).await
}

pub async fn require_verified(
    State(state): State<AppState>,
    mut req: Request<Body>,
    next: Next,
) -> Response {
    let cookie_header = req.headers().get("cookie").and_then(|v| v.to_str().ok());

    let session = match state.session_manager.get_session(cookie_header) {
        Ok(s) => s,
        Err(_) => return Redirect::to("/login").into_response(),
    };

    let user = match state.store.get_user_by_id(session.user_id).await {
        Ok(Some(u)) => u,
        _ => {
            let mut resp = Redirect::to("/login").into_response();
            if let Ok(clear) = state.session_manager.clear_session_cookie() {
                resp.headers_mut()
                    .insert(axum::http::header::SET_COOKIE, clear);
            }
            return resp;
        }
    };

    if user.verified_at.is_none() {
        return Redirect::to("/verify-pending").into_response();
    }

    req.extensions_mut().insert(AuthUserId(session.user_id));
    req.extensions_mut().insert(AuthSession(session));
    req.extensions_mut().insert(AuthUser(user));

    next.run(req).await
}
